File: src/main/java/com/freya02/botcommands/internal/components/sql/SQLFetchResult.java
package com.freya02.botcommands.internal.components.sql;

import com.freya02.botcommands.api.components.FetchResult;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.sql.Connection;

public class SQLFetchResult extends FetchResult {
	private final Connection connection;
	private boolean closed;

	public SQLFetchResult(@Nullable SQLFetchedComponent fetchedComponent, @NotNull Connection connection) {
		super(fetchedComponent);

		this.connection = connection;
	}

	public Connection getConnection() {
		if (closed) throw new IllegalStateException("Cannot get Connection from SQLFetchedComponent as it has been closed");

		return connection;
	}

	@Override
	@Nullable
	public SQLFetchedComponent getFetchedComponent() {
		return (SQLFetchedComponent) super.getFetchedComponent();
	}

	@Override
	public void close() throws Exception {
		closed = true;

		connection.close();
	}
}

File: src/main/kotlin/com/freya02/botcommands/internal/components/sql/SQLFetchResult.kt
package com.freya02.botcommands.internal.components.sql

import com.freya02.botcommands.api.components.FetchResult
import com.freya02.botcommands.internal.core.db.Transaction
import com.freya02.botcommands.internal.throwInternal

internal class SQLFetchResult(val fetchedComponent: SQLFetchedComponent?, transaction: Transaction?) : FetchResult(fetchedComponent) {
    private var closed = false

    private val _transaction = transaction
    val transaction: Transaction
        get() {
            check(!closed) { "Cannot get Transaction from SQLFetchedComponent as it has been closed" }
            return _transaction ?: throwInternal("Transaction was used when the fetch result was erroneous")
        }

    @Throws(Exception::class)
    override fun close() {
        closed = true
    }
}